Always use ClientJacksonMessageBodyReader over ServerJacksonMessageBody

Sgitario opened the pull request:
After https://github.com/quarkusio/quarkus/pull/27203, we can customize the object mappers to be used by REST Client Reactive. However, because of https://github.com/quarkusio/quarkus/pull/16368, the implementation was never picked up when the resteasy reactive Jackson extension is in place.  
To fix this issue, I tried to remove the ResteasyReactiveJacksonProviderDefinedBuildItem build item, and surprisingly everything kept working fine (I verified the test that was added as part of https://github.com/quarkusio/quarkus/pull/16368).

Fix https://github.com/quarkusio/quarkus/issues/23979

Sgitario commented on extensions/resteasy-reactive/rest-client-reactive-jackson/runtime/src/main/java/io/quarkus/rest/client/reactive/jackson/runtime/serialisers/ClientJacksonMessageBodyReader.java:
```diff
@@ -40,7 +39,7 @@ public Object readFrom(Class<Object> type, Type genericType, Annotation[] annota
             return super.readFrom(type, genericType, annotations, mediaType, httpHeaders, entityStream);
         } catch (StreamReadException | DatabindException e) {
-            throw new ClientWebApplicationException(e, Response.Status.BAD_REQUEST);
+            throw new WebApplicationException(e, Response.Status.BAD_REQUEST);
```
As now this message body ready is used by the client, this change is necessary, if not, the test BadRequestNotPropagatedTestCase fails. Though, I think this change is correct.

Sgitario commented:
Failures look related. Investigating.

gsmet commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java:
```diff
@@ -94,11 +94,6 @@ ServerDefaultProducesHandlerBuildItem jsonDefault() {
         return ServerDefaultProducesHandlerBuildItem.json();
     }
 
-    @BuildStep
-    ResteasyReactiveJacksonProviderDefinedBuildItem jacksonRegistered() {
-        return new ResteasyReactiveJacksonProviderDefinedBuildItem();
-    }
-
```
Let's keep the build item and continue to produce for now. Make it deprecated and we will remove it in a few versions.

(There might be external extensions relying on it)

Sgitario committed:
Always use ClientJacksonMessageBodyReader over ServerJacksonMessageBody

After https://github.com/quarkusio/quarkus/pull/27203, we can customize the object mappers to be used by REST Client Reactive.
However, because of https://github.com/quarkusio/quarkus/pull/16368, the implementation was never picked up when the resteasy reactive jackson extension is in place. 
I tried to remove the ResteasyReactiveJacksonProviderDefinedBuildItem build item and surprisingly everything kept working fine (I verified the test that was added as part of  https://github.com/quarkusio/quarkus/pull/16368.

Fix https://github.com/quarkusio/quarkus/issues/23979

Sgitario commented:
@gsmet , can you take another look into this please?

geoand commented:
@Sgitario I just saw this.

I prefer we close this one since I am dealing with it in https://github.com/quarkusio/quarkus/pull/30529

Sgitario commented:
> @Sgitario I just saw this.
> 
> I prefer we close this one since I am dealing with it in #30529

I think both fixes are complementary. Actually, the most important piece of this pull request is https://github.com/quarkusio/quarkus/pull/30436/files#diff-779f7936eaa23db9c900e262d5d37b9642ad56d2d58fa51dc6991b138d2a218aL45, if you're planning to incorporate it to your pull request, then I'm fine with it.

geoand commented:
Yes I know, just find that out from the various tests that were failing.

I would really prefer ro close this and keep the other one as I have already spent way too much time on the issue and would very much not like to have to deal with merge conflicts

Sgitario commented:
> Yes I know, just find that out from the various tests that were failing.
> 
> I would really prefer ro close this and keep the other one as I have already spent way too much time on the issue and would very much not like to have to deal with merge conflicts

Sure, as long as you add the mentioned line, anything that helps you most. Closing.

geoand commented:
👍🏼 

I have pushed an update (that still does not fix all the tests) that includes it.
